LoopKitUI/View Controllers/BasalScheduleTableViewController.swift

```diff
@@ -0,0 +1,396 @@
+
+        var startTime = TimeInterval(0)
+        var value = allowedBasalRates.count > 0 ? allowedBasalRates[0] : 0
+
+        if scheduleItems.count > 0 {
+            let lastItem = scheduleItems.last!
+
+            startTime = lastItem.startTime + minimumTimeInterval
+            value = lastItem.value
+
+            if startTime >= TimeInterval(hours: 24) {
+                return
+            }
+        }
+
+        scheduleItems.append(
+            RepeatingScheduleValue(
+                startTime: min(TimeInterval(hours: 23.5), startTime),
+                value: value
+            )
+        )
+        updateInsertButton()
+        updateTimeLimitsForItemsAdjacentTo(scheduleItems.count-1)
+
+        super.addScheduleItem(sender)
+    }
+
+    override func insertableIndiciesByRemovingRow(_ row: Int, withInterval timeInterval: TimeInterval) -> [Bool] {
+        return insertableIndices(for: scheduleItems, removing: row, with: timeInterval)
+    }
+
+    func preferredValueFractionDigits() -> Int {
+        return 1
+    }
+
+    public weak var syncSource: BasalScheduleTableViewControllerSyncSource? {
+        didSet {
+            isReadOnly = syncSource?.basalScheduleTableViewControllerIsReadOnly(self) ?? false
+
+            if isViewLoaded {
+                tableView.reloadData()
+            }
+        }
+    }
+
+    private var isSyncInProgress = false {
+        didSet {
+            for cell in tableView.visibleCells {
+                switch cell {
+                case let cell as TextButtonTableViewCell:
+                    cell.isEnabled = !isSyncInProgress
+                    cell.isLoading = isSyncInProgress
+                case let cell as BasalScheduleEntryTableViewCell:
+                    cell.isReadOnly = isReadOnly || isSyncInProgress
+                default:
+                    break
+                }
+            }
+
+            for item in navigationItem.rightBarButtonItems ?? [] {
+                item.isEnabled = !isSyncInProgress
+            }
+
+            navigationItem.hidesBackButton = isSyncInProgress
+        }
+    }
+
+    public var isScheduleValid: Bool {
+        for item in scheduleItems {
+            if !isBasalRateValid(item.value) {
+                return false
+            }
+        }
+        return scheduleItems.count <= maximumScheduleItemCount
+    }
+
+    private func updateTimeLimitsForItemsAdjacentTo(_ index: Int) {
+        let item = scheduleItems[index]
+
+        // Update previous cell's maximumStartTime
+        if index > 0 {
+            let previousIndexPath = IndexPath(row: index-1, section: Section.schedule.rawValue)
+            if let previousCell = tableView.cellForRow(at: previousIndexPath) as? BasalScheduleEntryTableViewCell {
+                previousCell.maximumStartTime = item.startTime - minimumTimeInterval
+            }
+        }
+        // Update next cell's minimumStartTime
+        if index < scheduleItems.count - 1 {
```

Comment:
Nit: some index checks in this file use `scheduleItems.count`, and some use `scheduleItems.endIndex`. I prefer the latter.